File: 543-diameter-of-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.max_depth = 0

        def max_depth(root):
            if root is None:
                return 0
            
            left = max_depth(root.left)
            right = max_depth(root.right)
            self.max_depth = max(self.max_depth, left + right)
            return 1 + max(left, right)
        
        max_depth(root)
        return self.max_depth

File: 543-diameter-of-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        postOrderVisit(root);
        return maxDepth;
    }

    int postOrderVisit(TreeNode * root) {
        if (root == nullptr) return 0;

        int leftDepth = postOrderVisit(root->left);
        int rightDepth = postOrderVisit(root->right);
        maxDepth = max(leftDepth + rightDepth, maxDepth);
        return 1 + max(leftDepth, rightDepth);
    }
};